src/Storages/ObjectStorage/StorageObjectStorageCluster.cpp

```diff
@@ -33,14 +57,18 @@ StorageObjectStorageCluster::StorageObjectStorageCluster(
     , object_storage(object_storage_)
 {
     ColumnsDescription columns{columns_};
-    resolveSchemaAndFormat(columns, configuration->format, object_storage, configuration, {}, context_);
+    std::string sample_path;
+    resolveSchemaAndFormat(columns, configuration->format, object_storage, configuration, {}, sample_path, context_);
     configuration->check(context_);
 
     StorageInMemoryMetadata metadata;
     metadata.setColumns(columns);
     metadata.setConstraints(constraints_);
 
-    setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(metadata.getColumns()));
+    if (sample_path.empty() && context_->getSettingsRef().use_hive_partitioning)
+        sample_path = getPathSample(metadata, context_);
+
+    setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(metadata.getColumns(), context_, sample_path, getFormatSettings(context_)));
```

Comment:
`getFormatSettings`  here and in similar places seems to be unnecessary, we do it in `getVirtualsForFileLikeStorage` if `format_settings = std::nullopt`